// seyfert/lib/events/hooks/soundboard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SOUNDBOARD_SOUNDS = exports.GUILD_SOUNDBOARD_SOUND_DELETE = exports.GUILD_SOUNDBOARD_SOUNDS_UPDATE = exports.GUILD_SOUNDBOARD_SOUND_UPDATE = exports.GUILD_SOUNDBOARD_SOUND_CREATE = void 0;
const client_1 = require("../../client");
const common_1 = require("../../common");
const GUILD_SOUNDBOARD_SOUND_CREATE = (self, data) => {
    return data.user ? { ...(0, common_1.toCamelCase)(data), user: client_1.Transformers.User(self, data.user) } : (0, common_1.toCamelCase)(data);
};
exports.GUILD_SOUNDBOARD_SOUND_CREATE = GUILD_SOUNDBOARD_SOUND_CREATE;
const GUILD_SOUNDBOARD_SOUND_UPDATE = (self, data) => {
    return data.user ? { ...(0, common_1.toCamelCase)(data), user: client_1.Transformers.User(self, data.user) } : (0, common_1.toCamelCase)(data);
};
exports.GUILD_SOUNDBOARD_SOUND_UPDATE = GUILD_SOUNDBOARD_SOUND_UPDATE;
const GUILD_SOUNDBOARD_SOUNDS_UPDATE = (self, data) => {
    return {
        guildId: data.guild_id,
        sounds: data.soundboard_sounds.map(d => d.user ? { ...(0, common_1.toCamelCase)(d), user: client_1.Transformers.User(self, d.user) } : (0, common_1.toCamelCase)(d)),
    };
};
exports.GUILD_SOUNDBOARD_SOUNDS_UPDATE = GUILD_SOUNDBOARD_SOUNDS_UPDATE;
const GUILD_SOUNDBOARD_SOUND_DELETE = (_, data) => {
    return (0, common_1.toCamelCase)(data);
};
exports.GUILD_SOUNDBOARD_SOUND_DELETE = GUILD_SOUNDBOARD_SOUND_DELETE;
const SOUNDBOARD_SOUNDS = (_, data) => {
    return (0, common_1.toCamelCase)(data);
};
exports.SOUNDBOARD_SOUNDS = SOUNDBOARD_SOUNDS;
