// seyfert/lib/client/transformers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transformers = void 0;
const commands_1 = require("../commands");
const structures_1 = require("../structures");
const Application_1 = require("../structures/Application");
exports.Transformers = {
    Application(...args) {
        return new Application_1.Application(...args);
    },
    ApplicationEmoji(...args) {
        return new structures_1.ApplicationEmoji(...args);
    },
    AnonymousGuild(...args) {
        return new structures_1.AnonymousGuild(...args);
    },
    AutoModerationRule(...args) {
        return new structures_1.AutoModerationRule(...args);
    },
    BaseChannel(...args) {
        return new structures_1.BaseChannel(...args);
    },
    BaseGuildChannel(...args) {
        return new structures_1.BaseGuildChannel(...args);
    },
    TextGuildChannel(...args) {
        return new structures_1.TextGuildChannel(...args);
    },
    DMChannel(...args) {
        return new structures_1.DMChannel(...args);
    },
    VoiceChannel(...args) {
        return new structures_1.VoiceChannel(...args);
    },
    StageChannel(...args) {
        return new structures_1.StageChannel(...args);
    },
    MediaChannel(...args) {
        return new structures_1.MediaChannel(...args);
    },
    ForumChannel(...args) {
        return new structures_1.ForumChannel(...args);
    },
    ThreadChannel(...args) {
        return new structures_1.ThreadChannel(...args);
    },
    CategoryChannel(...args) {
        return new structures_1.CategoryChannel(...args);
    },
    NewsChannel(...args) {
        return new structures_1.NewsChannel(...args);
    },
    DirectoryChannel(...args) {
        return new structures_1.DirectoryChannel(...args);
    },
    ClientUser(...args) {
        return new structures_1.ClientUser(...args);
    },
    Guild(...args) {
        return new structures_1.Guild(...args);
    },
    GuildBan(...args) {
        return new structures_1.GuildBan(...args);
    },
    Emoji(...args) {
        return new structures_1.Emoji(...args);
    },
    GuildEmoji(...args) {
        return new structures_1.GuildEmoji(...args);
    },
    GuildMember(...args) {
        return new structures_1.GuildMember(...args);
    },
    InteractionGuildMember(...args) {
        return new structures_1.InteractionGuildMember(...args);
    },
    GuildRole(...args) {
        return new structures_1.GuildRole(...args);
    },
    GuildTemplate(...args) {
        return new structures_1.GuildTemplate(...args);
    },
    Message(...args) {
        return new structures_1.Message(...args);
    },
    WebhookMessage(...args) {
        return new structures_1.WebhookMessage(...args);
    },
    Poll(...args) {
        return new structures_1.Poll(...args);
    },
    Sticker(...args) {
        return new structures_1.Sticker(...args);
    },
    User(...args) {
        return new structures_1.User(...args);
    },
    VoiceState(...args) {
        return new structures_1.VoiceState(...args);
    },
    Webhook(...args) {
        return new structures_1.Webhook(...args);
    },
    OptionResolver(...args) {
        return new commands_1.OptionResolver(...args);
    },
    Entitlement(...args) {
        return new structures_1.Entitlement(...args);
    },
};
