// seyfert/lib/cache/resources/bans.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bans = void 0;
const transformers_1 = require("../../client/transformers");
const common_1 = require("../../common");
const guild_based_1 = require("./default/guild-based");
class Bans extends guild_based_1.GuildBasedResource {
    namespace = 'ban';
    //@ts-expect-error
    filter(data, id, guild_id, from) {
        return true;
    }
    parse(data, key, guild_id) {
        const { user, ...rest } = super.parse(data, data.user?.id ?? key, guild_id);
        return rest;
    }
    get(id, guild) {
        return (0, common_1.fakePromise)(super.get(id, guild)).then(rawBan => rawBan ? transformers_1.Transformers.GuildBan(this.client, rawBan, guild) : undefined);
    }
    raw(id, guild) {
        return super.get(id, guild);
    }
    bulk(ids, guild) {
        return (0, common_1.fakePromise)(super.bulk(ids, guild)).then(bans => bans
            .map(rawBan => {
            return rawBan ? transformers_1.Transformers.GuildBan(this.client, rawBan, guild) : undefined;
        })
            .filter(x => x !== undefined));
    }
    bulkRaw(ids, guild) {
        return super.bulk(ids, guild);
    }
    values(guild) {
        return (0, common_1.fakePromise)(super.values(guild)).then(bans => bans
            .map(rawBan => {
            return rawBan ? transformers_1.Transformers.GuildBan(this.client, rawBan, guild) : undefined;
        })
            .filter(x => x !== undefined));
    }
    valuesRaw(guild) {
        return super.values(guild);
    }
}
exports.Bans = Bans;
