// seyfert/lib/cache/resources/channels.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Channels = void 0;
const common_1 = require("../../common");
const structures_1 = require("../../structures");
const guild_related_1 = require("./default/guild-related");
class Channels extends guild_related_1.GuildRelatedResource {
    namespace = 'channel';
    //@ts-expect-error
    filter(data, id, guild_id, from) {
        return true;
    }
    parse(data, id, guild_id) {
        const { permission_overwrites, ...rest } = super.parse(data, id, guild_id);
        return rest;
    }
    get(id) {
        return (0, common_1.fakePromise)(super.get(id)).then(rawChannel => rawChannel ? (0, structures_1.channelFrom)(rawChannel, this.client) : undefined);
    }
    raw(id) {
        return super.get(id);
    }
    bulk(ids) {
        return (0, common_1.fakePromise)(super.bulk(ids)).then(channels => channels.map(rawChannel => (0, structures_1.channelFrom)(rawChannel, this.client)));
    }
    bulkRaw(ids) {
        return super.bulk(ids);
    }
    values(guild) {
        return (0, common_1.fakePromise)(super.values(guild)).then(channels => channels.map(rawChannel => (0, structures_1.channelFrom)(rawChannel, this.client)));
    }
    valuesRaw(guild) {
        return super.values(guild);
    }
}
exports.Channels = Channels;
