// seyfert/lib/cache/resources/voice-states.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoiceStates = void 0;
const transformers_1 = require("../../client/transformers");
const common_1 = require("../../common");
const guild_based_1 = require("./default/guild-based");
class VoiceStates extends guild_based_1.GuildBasedResource {
    namespace = 'voice_state';
    //@ts-expect-error
    filter(data, id, guild_id, from) {
        return true;
    }
    parse(data, id, guild_id) {
        const { member, ...rest } = super.parse(data, id, guild_id);
        return rest;
    }
    get(memberId, guildId) {
        return (0, common_1.fakePromise)(super.get(memberId, guildId)).then(state => state ? transformers_1.Transformers.VoiceState(this.client, state) : undefined);
    }
    raw(memberId, guildId) {
        return super.get(memberId, guildId);
    }
    bulk(ids, guild) {
        return (0, common_1.fakePromise)(super.bulk(ids, guild)).then(states => states
            .map(state => (state ? transformers_1.Transformers.VoiceState(this.client, state) : undefined))
            .filter(x => x !== undefined));
    }
    bulkRaw(ids, guild) {
        return super.bulk(ids, guild);
    }
    values(guildId) {
        return (0, common_1.fakePromise)(super.values(guildId)).then(states => states.map(state => transformers_1.Transformers.VoiceState(this.client, state)));
    }
    valuesRaw(guildId) {
        return super.values(guildId);
    }
}
exports.VoiceStates = VoiceStates;
