// seyfert/lib/common/shorters/soundboard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SoundboardShorter = void 0;
const base_1 = require("./base");
class SoundboardShorter extends base_1.BaseShorter {
    getDefaults() {
        return this.client.proxy['soundboard-default-sounds'].get();
    }
    send(channelId, body) {
        return this.client.proxy.channels(channelId)['send-soundboard-sound'].post({
            body,
        });
    }
    list(guildId) {
        return this.client.proxy.guilds(guildId)['soundboard-sounds'].get();
    }
    get(guildId, soundID) {
        return this.client.proxy.guilds(guildId)['soundboard-sounds'](soundID).get();
    }
    create(guildId, body) {
        return this.client.proxy.guilds(guildId)['soundboard-sounds'].post({
            body,
        });
    }
    edit(guildId, soundID, body) {
        return this.client.proxy.guilds(guildId)['soundboard-sounds'](soundID).patch({
            body,
        });
    }
    delete(guildId, soundID, reason) {
        return this.client.proxy.guilds(guildId)['soundboard-sounds'](soundID).delete({ reason });
    }
}
exports.SoundboardShorter = SoundboardShorter;
