// seyfert/lib/commands/applications/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStringOption = createStringOption;
exports.createIntegerOption = createIntegerOption;
exports.createNumberOption = createNumberOption;
exports.createChannelOption = createChannelOption;
exports.createBooleanOption = createBooleanOption;
exports.createUserOption = createUserOption;
exports.createRoleOption = createRoleOption;
exports.createMentionableOption = createMentionableOption;
exports.createAttachmentOption = createAttachmentOption;
exports.createMiddleware = createMiddleware;
const types_1 = require("../../types");
function createStringOption(data) {
    return { ...data, type: types_1.ApplicationCommandOptionType.String };
}
function createIntegerOption(data) {
    return { ...data, type: types_1.ApplicationCommandOptionType.Integer };
}
function createNumberOption(data) {
    return { ...data, type: types_1.ApplicationCommandOptionType.Number };
}
function createChannelOption(data) {
    return { ...data, type: types_1.ApplicationCommandOptionType.Channel };
}
function createBooleanOption(data) {
    return { ...data, type: types_1.ApplicationCommandOptionType.Boolean };
}
function createUserOption(data) {
    return { ...data, type: types_1.ApplicationCommandOptionType.User };
}
function createRoleOption(data) {
    return { ...data, type: types_1.ApplicationCommandOptionType.Role };
}
function createMentionableOption(data) {
    return { ...data, type: types_1.ApplicationCommandOptionType.Mentionable };
}
function createAttachmentOption(data) {
    return { ...data, type: types_1.ApplicationCommandOptionType.Attachment };
}
function createMiddleware(data) {
    return data;
}
