// seyfert/lib/structures/Application.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Application = void 0;
const DiscordBase_1 = require("./extra/DiscordBase");
/**
 * Due to current limitations, this is exclusively for the current application.
 */
class Application extends DiscordBase_1.DiscordBase {
    constructor(client, data) {
        // override any id for safety
        data.id = client.applicationId;
        super(client, data);
    }
    /**
     * Fetch the current application.
     */
    fetch() {
        return this.client.applications.fetch();
    }
    /**
     * Edit the current application.
     */
    edit(data) {
        return this.client.applications.edit(data);
    }
    /**
     * Get an activity instance.
     */
    getActivityInstance(instanceId) {
        return this.client.applications.getActivityInstance(instanceId);
    }
    emojis = {
        /**
         * Get an application emoji.
         */
        fetch: (id) => this.client.applications.getEmoji(id),
        /**
         * Get the application emojis.
         */
        list: () => this.client.applications.listEmojis(),
        /**
         * Create an application emoji.
         */
        create: (data) => this.client.applications.createEmoji(data),
        /**
         * Edit an application emoji.
         */
        edit: (emojiId, body) => this.client.applications.editEmoji(emojiId, body),
    };
}
exports.Application = Application;
