// seyfert/lib/types/payloads/webhokEvents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookEventTypes = exports.WebhookRequestType = void 0;
/**
 * https://discord.com/developers/docs/events/webhook-events#webhook-types
 */
var WebhookRequestType;
(function (WebhookRequestType) {
    /** PING event sent to verify your Webhook Event URL is active */
    WebhookRequestType[WebhookRequestType["PING"] = 0] = "PING";
    /** Webhook event (details for event in event body object) */
    WebhookRequestType[WebhookRequestType["Event"] = 1] = "Event";
})(WebhookRequestType || (exports.WebhookRequestType = WebhookRequestType = {}));
/**
 * https://discord.com/developers/docs/events/webhook-events#event-types
 */
var WebhookEventTypes;
(function (WebhookEventTypes) {
    /** Sent when an app was authorized by a user to a server or their account */
    WebhookEventTypes["ApplicationAuthorized"] = "APPLICATION_AUTHORIZED";
    /** Entitlement was created */
    WebhookEventTypes["EntitlementCreate"] = "ENTITLEMENT_CREATE";
    /**
     * User was added to a Quest (currently unavailable)
     * @unstable
     */
    WebhookEventTypes["QuestUserEnrollment"] = "QUEST_USER_ENROLLMENT";
})(WebhookEventTypes || (exports.WebhookEventTypes = WebhookEventTypes = {}));
